### ISSUE TITLE
Handling variable minibatch sizes with sequence inputs in reporter, optimizer

### ISSUE THREAD

**Issue by hknerdgn** [chainer/chainer#4004]
In some problems with sequential input, a minibatch contains a fixed number of sequences of differing lengths/durations/time-steps. Total number of time steps is variable from minibatch to minibatch. Most loss functions that depend on a single output at each time step need to be weighted to obtain a correct overall loss function (and not simply taking their mean over minibatches). Currently, chainer's reporter uses a Summary class that does not take into account possible weighting of each minibatch (due to variable sizes). This can be remedied by taking optionally weighted averages for the reporter's Summary class. When adding variables to the reporter, we can add a tuple of (value, weight) which can do weighting of each value.

In addition, it helps to adjust the learning rate for variable minibatch size updates. In any case, making the learning rate proportional to minibatch size is useful. This feature(s) can be added to chainer as well. May be optimizer.update() can be made to accept a weight/scale factor or it can be handled internally as interpreting the learning rate as a learning rate per sample (or time step) instead of a fixed learning rate for each minibatch.
